Updates scan for Signatures to use filepath.Walk()

NandoTheessen opened the pull request:
Hi 👋🏻 

I'm fairly new to go and wanted to make an attempt to contribute. 
I've started with the findGoSigs function, if the approach is correct I'll port the RegoSig function as well. 

Looking for any feedback and or pointers 👍🏻 

Addresses #501

itaysk commented on tracee-rules/signature.go:
```diff
@@ -45,27 +45,28 @@ func getSignatures(rulesDir string, rules []string) ([]types.Signature, error) {
-			continue
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("error opening plugin %s: %v", info.Name(), err)
```
to my understanding this will cause the walker to abort. isn't it better to return SkipDir (and log it) so that other files might be still processed? same goes to all other errors returned from this func

itaysk replied:
didn't you mean to SkipDir here too?

itaysk replied:
also, log the error like in other cases

itaysk commented on tracee-rules/signature.go:
```diff
@@ -45,27 +45,28 @@ func getSignatures(rulesDir string, rules []string) ([]types.Signature, error) {
+		if err != nil {
+			return fmt.Errorf("error opening plugin %s: %v", info.Name(), err)
+		}
+		if filepath.Ext(info.Name()) != ".so" {
```
please add a check that this is a "regular" file. so that directories ending with ".so" won't be processed.

itaysk reviewed (commented):
thanks @NandoTheessen !
other than the comments inline, please also update the help to of the --rules-dir to indicate that this is a recursive search

itaysk commented:
oh and also, feel free to include the rego rules implementation in the same PR, it wouldn't make much sense to change just one and not the other

itaysk commented:
I'm also curious, what does `*upd:` means in the commit message?

NandoTheessen commented:
Thanks for the pointers itaysk 👋🏻
I've implemented the changes you requested, but wasn't sure on the wording for the readme file.
Please lmk if you'd like it differently.

I usually contain *upd (update) *fix or *merge in my commit messages to provide a context what kind of change I'm committing, I've changed that for the following commit messages, since I don't know how you handle it for this project.

itaysk commented on tracee-rules/signature.go:
```diff
@@ -45,52 +45,55 @@ func getSignatures(rulesDir string, rules []string) ([]types.Signature, error) {
 		}
 		sigs := *export.(*[]types.Signature)
 		res = append(res, sigs...)
+		return filepath.SkipDir
```
is this necessary?

NandoTheessen replied:
No, it's not. I've changed it to return `nil` instead.

itaysk commented on tracee-rules/Readme.md:
```diff
@@ -34,7 +34,7 @@ This will:
 
 # Rules
-Rules are discovered from the local `rules` directory (unless changed by the `--rules-dir` flag). By default, all discovered rules will be loaded unless specific rules are selected using the `--rules` flag.
+Rules are discovered from the local `rules` directory and all it's subfolders (unless changed by the `--rules-dir` flag). By default, all discovered rules will be loaded unless specific rules are selected using the `--rules` flag.
```
nit: change subfolders to sub-directories so that it's consistent with the beginning of the sentence

simar7 commented:
Thanks @NandoTheessen this is looking good! Could you also add a few basic tests to test this new functionality? I read that you are new to Go so here's a few tips to write your first test:

1. A simple test in Go would look like this:
```
func TestFoo(t *testing.T){
    want := "foo"
    got := fooFunc(input)
    assert.Equal(t, got, want)
}
```
to make a simple assert that `want == got` is true. 

2. In this PR you can call the `findGoSigs` func and the `findRegoSigs` func with `dir` as the input argument. In Go testing you can create a temporary directory just for a test and populate it with your test artifacts. See https://golang.org/pkg/io/ioutil/#TempDir and https://golang.org/pkg/io/ioutil/#TempFile for details.
3. When you have several input combinations, you can convert them into a table to form a table driven test. Here's an example of it in action https://github.com/aquasecurity/tracee/blob/main/tracee-rules/input_test.go

Let us know if you need help writing tests!

simar7 commented on tracee-rules/signature.go:
```diff
@@ -45,63 +44,62 @@ func getSignatures(rulesDir string, rules []string) ([]types.Signature, error) {
-		if file.Name() == "helpers.rego" {
-			continue
+	var res []types.Signature
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
```
The `err` passed into `filepath.Walk` must be checked before the actual function logic is executed. If a non nil error is passed in, the resulting info will be `nil` and this function will panic.

See another example here: https://github.com/aquasecurity/trivy/pull/208


Details: https://golang.org/pkg/path/filepath/#WalkFunc

simar7 commented on tracee-rules/signature.go:
```diff
@@ -45,63 +44,62 @@ func getSignatures(rulesDir string, rules []string) ([]types.Signature, error) {
-			continue
+	var res []types.Signature
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if filepath.Ext(info.Name()) != ".rego" || info.IsDir() == true || info.Name() == "helpers.rego" {
+			return filepath.SkipDir
 		}
```
This isn't an error but rather a `return nil` instead. Returning an error here will completely skip this directory. I found this out by writing a simple test:

```
func TestFindRegoSigs(t *testing.T) {
	actual, err := findRegoSigs("signatures/rego")
	require.NoError(t, err)
	assert.NotEmpty(t, actual)
}
```

simar7 reviewed (changes requested):
Left some comments.

simar7 commented:
Are you still interested in making those changes @NandoTheessen? Please let us know.

NandoTheessen commented:
Yes, been a little busy but would like to finish this up on the weekend :)

yanivagman commented:
As per @simar7 comment in #501 

> Since the creation of this issue, Go 1.16 has released filepath.WalkDir which guarantees better perf than filepath.Walk. https://golang.org/pkg/path/filepath/#WalkDir

Should we keep this PR open? Also, there were no changes made here for the last 5 months

itaysk commented:
looking at the feedback on this PR:
1. no tests - it would have been nice if the contributor added tests, but since the function was originally untested I wouldn't fight on this.
2. newer API slightly more efficient - remember this is small piece of code that is running once at init and is doing very simple work. I don't think it matters which API we use.
3. some code review comment - this IS a blocker. so let give it a couple of more days and if no response let's close.
